src/laser_core/migration.py

```diff
@@ -0,0 +1,394 @@
+on the Earth's surface using the Haversine formula.
+
+Functions:
+    gravity(pops: np.ndarray, distances: np.ndarray, k: float, a: float, b: float, c: float, max_frac: Union[float, None]=None, **kwargs) -> np.ndarray:
+
+    row_normalizer(network: np.ndarray, max_rowsum: float) -> np.ndarray:
+        Normalize the rows of a given network matrix such that no row sum exceeds a specified maximum value.
+
+    competing_destinations(pops: np.ndarray, distances: np.ndarray, b: float, c: float, delta: float, **params) -> np.ndarray:
+
+    stouffer(pops: np.ndarray, distances: np.ndarray, k: float, a: float, b: float, include_home: bool, **params) -> np.ndarray:
+        Compute a migration network using a modified Stouffer's model.
+
+    radiation(pops: np.ndarray, distances: np.ndarray, k: float, include_home: bool, **params) -> np.ndarray:
+
+    distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
+        Calculate the great-circle distance between two points on the Earth's surface using the Haversine formula.
+"""
+
+from numbers import Number
+
+import numpy as np
+
+
+def shared_sanity_checks(pops, distances, **params):
+    _is_instance(pops, np.ndarray, f"pops must be a NumPy array ({type(pops)=})")
+    _has_dimensions(pops, 1, f"pops must be a 1D array ({pops.shape=})")
+    _is_dtype(pops, np.number, f"pops must be a numeric array ({pops.dtype=})")
+    _has_values(pops >= 0, "pops must contain only non-negative values")
+
+    _is_instance(distances, np.ndarray, f"distances must be a NumPy array ({type(distances)=})")
+    _has_dimensions(distances, 2, f"distances must be a 2D array ({distances.shape=})")
+    _has_shape(
+        distances,
+        (pops.shape[0], pops.shape[0]),
+        f"distances must be a square matrix with length equal to the length of pops ({distances.shape=}, {pops.shape=})",
+    )
+    _is_dtype(distances, np.number, f"distances must be a numeric array ({distances.dtype=})")
+    _has_values(distances >= 0, "distances must contain only non-negative values")
+    _has_values(distances == distances.T, "distances must be a symmetric matrix")
+
+    a = params.get("a", None)
+    if a is not None:
+        _is_instance(a, Number, f"a must be a numeric value ({type(a)=})")
+        _has_values(a >= 0, f"a must be a non-negative value ({a=})")
+
+    b = params.get("b", None)
+    if b is not None:
+        _is_instance(b, Number, f"b must be a numeric value ({type(b)=})")
+        _has_values(b >= 0, f"b must be a non-negative value ({b=})")
+
+    c = params.get("c", None)
+    if c is not None:
+        _is_instance(c, Number, f"c must be a numeric value ({type(c)=})")
+        _has_values(c >= 0, f"c must be a non-negative value ({c=})")
+
+    delta = params.get("delta", None)
+    if delta is not None:
+        _is_instance(delta, Number, f"delta must be a numeric value ({type(delta)=})")
+
+    k = params.get("k", None)
+    if k is not None:
```

Comment:
Code coverage says we aren't ever testing `k`. We should have test cases which pass bad `k` types and values.